control-plane/agents/src/bin/core/volume/snapshot_operations.rs

```diff
@@ -285,31 +287,42 @@ impl ResourcePruning for OperationGuardArc<VolumeSnapshot> {
 }
 
 impl OperationGuardArc<VolumeSnapshot> {
-    fn snapshot_params(&self, replica: &ChildItem) -> Result<PrepareVolumeSnapshot, SvcError> {
+    fn snapshot_params(
+        &self,
+        replicas: &Vec<ChildItem>,
+    ) -> Result<PrepareVolumeSnapshot, SvcError> {
         let Some(parameters) = self.as_ref().prepare() else {
             return Err(SvcError::AlreadyExists {
                 id: self.uuid().to_string(),
                 kind: ResourceKind::VolumeSnapshot,
             });
         };
+        let mut repica_states = vec![];
+        let mut replica_snapshots = vec![];
+
         let volume = self.as_ref().spec().source_id();
         let generic_params = parameters.params().clone();
-        let snapshot_source = ReplicaSnapshotSource::new(
-            replica.spec().uid().clone(),
-            replica.state().pool_id.clone(),
-            replica.state().pool_uuid.clone().unwrap_or_default(),
-        );
-        let replica_snapshot = ReplicaSnapshot::new_vol(
-            ReplicaSnapshotSpec::new(&snapshot_source, SnapshotId::new()),
-            SnapshotParameters::new(volume, generic_params),
-            replica.state().size,
-            0,
-            replica.spec().size,
-        );
-        let replica = replica.state().clone();
+        for replica in replicas {
+            let snapshot_source = ReplicaSnapshotSource::new(
+                replica.spec().uid().clone(),
+                replica.state().pool_id.clone(),
+                replica.state().pool_uuid.clone().unwrap_or_default(),
+            );
+            let replica_snapshot = ReplicaSnapshot::new_vol(
+                ReplicaSnapshotSpec::new(&snapshot_source, SnapshotId::new()),
+                SnapshotParameters::new(volume, generic_params.clone()),
+                replica.state().size,
+                0,
+                replica.spec().size,
+            );
+            replica_snapshots.push(replica_snapshot);
+            let replica = replica.state().clone();
+            repica_states.push(replica);
+        }
+
         Ok(PrepareVolumeSnapshot {
             parameters,
-            replica_snapshot: (replica, replica_snapshot),
+            replica_snapshot: (repica_states, replica_snapshots),
```

Comment:
hmm is this the correct way, or should it be a vec of (state, snapshot)? Then we already have our list matched correctly?

Reply:
It is one to one mapping from replica to replica_snapshot. But I am thinking better to use a map where key is replica and value is replica_snapshot. It will give a better clarity.